bgpd/bgp_vty.c

```diff
@@ -11325,6 +11869,166 @@ void bgp_config_write_redistribute(struct vty *vty, struct bgp *bgp, afi_t afi,
+				    bgp_get_default(), bgp);
+	}
+	return CMD_SUCCESS;
+}
+
+DEFUN (bgp_no_export_vpn,
+       bgp_no_export_vpn_cmd,
+       "no export vpn",
+       NO_STR
+       "Export routes to another routing protocol\n"
+       "to VPN RIB per vpn-policy")
+{
+	VTY_DECLVAR_CONTEXT(bgp, bgp);
+	int was_on = 0;
+	afi_t afi;
+	safi_t safi;
+
+	if (BGP_INSTANCE_TYPE_VRF != bgp->inst_type
+	    && BGP_INSTANCE_TYPE_DEFAULT != bgp->inst_type) {
+		vty_out(vty,
+			"%% export vpn valid only for bgp vrf or default instance\n");
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+	afi = bgp_node_afi(vty);
+	safi = bgp_node_safi(vty);
+	if ((SAFI_UNICAST != safi) || ((AFI_IP != afi) && (AFI_IP6 != afi))) {
+		vty_out(vty,
+			"%% export vpn valid only for unicast ipv4|ipv6\n");
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	if (CHECK_FLAG(bgp->af_flags[afi][safi],
+		       BGP_CONFIG_VRF_TO_MPLSVPN_EXPORT)) {
+		was_on = 1;
+	}
+	if (was_on) {
+		/* trigger un-export current vrf */
+		zlog_debug("%s: calling postchange", __func__);
+		vpn_leak_prechange(BGP_VPN_POLICY_DIR_TOVPN, afi,
+				   bgp_get_default(), bgp);
+	}
+	UNSET_FLAG(bgp->af_flags[afi][safi], BGP_CONFIG_VRF_TO_MPLSVPN_EXPORT);
+	return CMD_SUCCESS;
+}
+
+static void bgp_vpn_policy_config_write_afi(struct vty *vty, struct bgp *bgp,
+					    afi_t afi)
+{
+	vty_frame(vty, " vpn-policy ipv%d\n", ((afi == AFI_IP) ? 4 : 6));
+
+	if (CHECK_FLAG(bgp->vpn_policy[afi].flags,
+		       BGP_VPN_POLICY_TOVPN_LABEL_SET)) {
+		vty_out(vty, "    label %u\n",
+			bgp->vpn_policy[afi].tovpn_label);
+	}
+	if (CHECK_FLAG(bgp->vpn_policy[afi].flags,
+		       BGP_VPN_POLICY_TOVPN_RD_SET)) {
+		char buf[RD_ADDRSTRLEN];
+		vty_out(vty, "    rd %s\n",
+			prefix_rd2str(&bgp->vpn_policy[afi].tovpn_rd, buf,
+				      sizeof(buf)));
+	}
+	if (CHECK_FLAG(bgp->vpn_policy[afi].flags,
+		       BGP_VPN_POLICY_TOVPN_NEXTHOP_SET)) {
+
+		char buf[PREFIX_STRLEN];
+		if (inet_ntop(bgp->vpn_policy[afi].tovpn_nexthop.family,
+			      &bgp->vpn_policy[afi].tovpn_nexthop.u.prefix, buf,
+			      sizeof(buf))) {
+
+			vty_out(vty, "    nexthop %s\n", buf);
+		}
+	}
+	if (bgp->vpn_policy[afi].rtlist[BGP_VPN_POLICY_DIR_FROMVPN]
+	    && bgp->vpn_policy[afi].rtlist[BGP_VPN_POLICY_DIR_TOVPN]
+	    && ecommunity_cmp(
+		       bgp->vpn_policy[afi].rtlist[BGP_VPN_POLICY_DIR_FROMVPN],
+		       bgp->vpn_policy[afi].rtlist[BGP_VPN_POLICY_DIR_TOVPN])) {
+
+		char *b = ecommunity_ecom2str(
+			bgp->vpn_policy[afi].rtlist[BGP_VPN_POLICY_DIR_TOVPN],
+			ECOMMUNITY_FORMAT_ROUTE_MAP, ECOMMUNITY_ROUTE_TARGET);
+		vty_out(vty, "    rt both %s\n", b);
+		XFREE(MTYPE_ECOMMUNITY_STR, b);
+	} else {
+		if (bgp->vpn_policy[afi].rtlist[BGP_VPN_POLICY_DIR_FROMVPN]) {
+			char *b = ecommunity_ecom2str(
+				bgp->vpn_policy[afi]
+					.rtlist[BGP_VPN_POLICY_DIR_FROMVPN],
+				ECOMMUNITY_FORMAT_ROUTE_MAP,
+				ECOMMUNITY_ROUTE_TARGET);
+			vty_out(vty, "    rt fromvpn %s\n", b);
+			XFREE(MTYPE_ECOMMUNITY_STR, b);
```

Comment:
was there an agreement on the vty naming convention ? 
rt from vpn <=> rt export
rt to vpn <=> rt import
 ?

Reply:
I do not think we achieved consensus. This is placeholder CLI so we can exercise code until we finalize syntax.